resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala

```diff
@@ -111,66 +109,104 @@ private[spark] class Client(
             DRIVER_LAUNCHER_SELECTOR_LABEL -> driverLauncherSelectorValue,
             SPARK_APP_NAME_LABEL -> appName)
           ++ parsedCustomLabels).asJava
-        val (servicePorts, containerPorts) = configurePorts()
-        val service = kubernetesClient.services().createNew()
-          .withNewMetadata()
-            .withName(kubernetesAppId)
-            .withLabels(Map(SPARK_APP_NAME_LABEL -> appName).asJava)
-            .endMetadata()
-          .withNewSpec()
-            .withSelector(resolvedSelectors)
-            .withPorts(servicePorts.asJava)
-            .endSpec()
-          .done()
-        sparkConf.set("spark.kubernetes.driver.service.name", service.getMetadata.getName)
-        sparkConf.setIfMissing("spark.driver.port", DEFAULT_DRIVER_PORT.toString)
-        sparkConf.setIfMissing("spark.blockmanager.port", DEFAULT_BLOCKMANAGER_PORT.toString)
-        val submitRequest = buildSubmissionRequest()
+        val containerPorts = configureContainerPorts()
         val submitCompletedFuture = SettableFuture.create[Boolean]
         val secretDirectory = s"$SPARK_SUBMISSION_SECRET_BASE_DIR/$kubernetesAppId"
-
+        val submitPending = new AtomicBoolean(false)
```

Comment:
More that I'm not sure if the implementation of watches can cause this method to be running twice simultaneously. If this method could be run again while a first invocation of the method has not completed and hence not set the future's status, then yes it would be possible that the upload logic is triggered twice.